File: src/latest.py
#!/usr/bin/env python
import boto3, json, os, os.path as op
from boto3.dynamodb.conditions import Key

TableName = 'cubelog'

ddb = boto3.resource('dynamodb')
table = ddb.Table(TableName)

last12 = table.query(KeyConditionExpression=Key('puzzle').eq('3x3x3'),
                     ScanIndexForward=False, Limit=12)['Items']
for n in [12, 5]:
  total = 0
  min_solve = float('inf')
  max_solve = 0
  count = 0
  for item in reversed(last12[0:n]):
    count = count + 1
    if n == 12:
      print(f'{item["timestamp"]}: {item["solve_time"]} ({item["scramble"]})')
    solve = float(item['solve_time'])
    total += solve
    if solve < min_solve:
        min_solve = solve
    if solve > max_solve:
        max_solve = solve

  avg = (total - min_solve - max_solve)/(count - 2)
  print(f'\nAverage of last {n}: {avg:.2f}')

File: src/latest.pl
#!/usr/bin/env perl
use v5.16;
use JSON;
use Paws::DynamoDB;

our $TableName = 'cubelog';

my $ddb = Paws->service('DynamoDB', region => $ENV{AWS_DEFAULT_REGION} // 'us-east-1');

my $result = $ddb->Query(TableName => $TableName,
                         ExpressionAttributeValues => { ":3x3" => { S => '3x3x3' } },
                         KeyConditionExpression => 'puzzle = :3x3',
                         ScanIndexForward => 0, Limit => 12);

my @last12 = @{$result->Items};

foreach my $n (12, 5) {
  my $total = 0;
  my $min_solve = 'inf';
  my $max_solve = '-inf';
  my $count = 0;
  foreach my $item (reverse @last12[0..$n-1]) {
    my %data = (map { $_ => $item->{Map}{$_}{S} } keys %{$item->{Map}});
    $count++;
    say "$data{timestamp}: $data{solve_time} ($data{scramble})" if $n == 12;
    $total += (my $solve = $data{solve_time});
    $min_solve = $solve if $solve < $min_solve;
    $max_solve = $solve if $solve > $max_solve;
  }

  my $avg = ($total - $min_solve - $max_solve)/($count - 2);
  printf "\nAverage of last $n: %.2f\n", $avg;
}
